### PR TITLE
WIP V0.2.0 - New Theme System

### PR DESCRIPTION
**Pull request by mattapperson**
New in this PR...
- [x]  All new theme system and API for theming based on StyleSheet API
- [x]   New Carousel component that now snaps and no scroll bar on web and autoplay
- [x]   Moved to storybook for docs as RSG is no longer maintained
- [x]   Moved to expo for native example app
- [x]   Co-located docs, tests, and components
- [x]   Updated RN, and React
- [x]   Refreshed default theme colors
- [x]   Merged TabButton into SegmentedControlButton
- [x]   use font icons

Documentation for the theme system can be found here: https://github.com/mattapperson/nachos-ui/blob/new-theme/src/Theme/ThemeProvider.md

### CODE REVIEW

File: src/Theme/index.js [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+    themeConfig = JSON.parse(
+      JSON.stringify(themeConfig).replace(/@([\w_-]+)/gm, (match, key) => {
+        if (!this.branding[key]) console.log(key);
+
+        return this.branding[key];
+      })
+    );
+
+    return (this.themeConfigs[name] = {
```
Comment (petrbrzek) [inline src/Theme/index.js]: This part is a little bit hard to follow.

File: src/Theme/index.js [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+              />
+            );
+          }}
+        </ThemeContext.Consumer>
+      );
+    }
+  };
+}
+
+function isObject(item) {
```
Comment (petrbrzek) [inline src/Theme/index.js]: Should not we test those utils? Or use specific dependencies like https://github.com/sindresorhus/is-obj (basically everything is made by sindresorhus 😀)

File: src/Theme/index.js [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+});
+
+export class Provider extends React.Component {
+  themeConfigs = {};
+  branding = defaultBranding;
+
+  constructor(props) {
+    super(props);
+    if (props.branding) {
+      this.branding = deepMerge(this.branding, props.branding);
```
Comment (petrbrzek) [inline src/Theme/index.js]: There is no deepMerge func.

Comment (mattapperson) [reply src/Theme/index.js]: HA! I guess I should also add some tests for this

**Comment by petrbrzek** [comment]
I'll continue tomorrow. Good night 🌚

File: src/Theme/index.js [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+              />
+            );
+          }}
+        </ThemeContext.Consumer>
+      );
+    }
+  };
+}
+
+function mergeDeep(target, ...sources) {
```
Comment (petrbrzek) [inline src/Theme/index.js]: This as well should be 1) tested and probably moved to utils 2) or we should use specialized dep.

File: src/Theme/index.js [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+  }
+
+  updateThemeConfig = (name, themeConfig) => {
+    if (this.props.theme && this.props.theme[name])
+      themeConfig = mergeDeep(themeConfig, this.props.theme[name]);
+
+    // Turn the themeConfig object into a string, then replace and strings that start with @ with a value from the branding config
+    // so `@darkColor` would pull the value of `this.branding.darkColor`
+    // We then convert the result back into an object
+    themeConfig = JSON.parse(
```
Comment (petrbrzek) [inline src/Theme/index.js]: I was thinking if this is the "best" JavaScript way to solve this. I'm open to discussion. What if the API was like this:

```
class A extends Component {
  static themeConfig = ({ linkColor }) => {
    style: {
      base: {
        fontSize: 14,
        color: linkColor,
      }
    }
  };
}
```

What do you think of this?

**Comment by petrbrzek** [comment]
Am I understand it correctly that storybook is going to be used instead of https://avocode.com/nachos-ui/docs/#!.?

File: src/Theme/index.js (hunk 5 above)
Comment (mattapperson) [reply src/Theme/index.js]: I thought about doing this. The "issue" here is that there is no ability to ensure variables are used correctly. Thus values could be `undefined` and hard to track down.
using a templeted string has the downside of being "less javascripty" and also less flexible... but we can add tests to confirm correct values are used... open to feedback on this though

**Comment by mattapperson** [comment]
> Am I understand it correctly that storybook is going to be used instead of https://avocode.com/nachos-ui/docs/#!.?

Correct, it seems the best alternative for a still-maintained docs provider?

File: src/Theme/index.js (hunk 4 above)
Comment (mattapperson) [reply src/Theme/index.js]: @petrbrzek I agree... however before doing this I would kinda like to see this PR merged and released as a beta version to ensure all of this is even in the right direction before locking in too much? Thoughts?
